### PR TITLE
add new nop command features

### PR DESCRIPTION
**Pull request by therealdreg**
## Description/Motivation/Screenshots

```
The `nop` command allows you to easily patch instructions with nops.

```
nop [LOCATION] [--i ITEMS] [--f] [--n] [--b]
```

`LOCATION` address/symbol to patch (by default this command replaces whole instructions)

`--i ITEMS` number of items to insert (default 1)

`--f` Force patch even when the selected settings could overwrite partial instructions

`--n` Instead of replacing whole instructions, insert ITEMS nop instructions, no matter how many instructions it overwrites

`--b` Instead of replacing whole instructions, fill ITEMS bytes with nops

gef➤ 	nop
gef➤ 	nop $pc+3
gef➤ 	nop --i 2 $pc+3
gef➤ 	nop --b
gef➤ 	nop --b $pc+3
gef➤ 	nop --f --b --i 2 $pc+3
gef➤ 	nop --n --i 2 $pc+3

```

## Against which architecture was this tested ?

"Tested" indicates that the PR works *and* the unit test (see `docs/testing.md`) run passes without issue.

 * [x] x86-32
 * [x] x86-64
 * [ ] ARM
 * [ ] AARCH64
 * [ ] MIPS
 * [ ] POWERPC
 * [ ] SPARC
 * [ ] RISC-V

## Checklist

- [x] My PR was done against the `dev` branch, not `main`.
- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [x] If my change adds new code, [adequate tests](docs/testing.md) have been added.
- [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

File: docs/commands/nop.md [hunk 1]
```diff
@@ -3,21 +3,26 @@
 The `nop` command allows you to easily patch instructions with nops.
 
 ```
-nop [LOCATION] [--n NUM_ITEMS] [--b]
+nop [LOCATION] [--i ITEMS] [--f] [--n] [--b]
 ```
 
-`LOCATION` address/symbol to patch
+`LOCATION` address/symbol to patch (by default this command replace whole instruction(s))
```
Comment (Grazfather) [inline docs/commands/nop.md:9]:
```suggestion
`LOCATION` address/symbol to patch (by default this command replaces whole instructions)
```

File: docs/commands/nop.md [hunk 2]
```diff
@@ -3,21 +3,26 @@
...
 
-`--n NUM_ITEMS` Instead of writing one instruction/nop, patch the specified number of
-instructions/nops (full instruction size by default)
+`--i ITEMS` number of items to insert (default 1)
+
+`--f` Force patch when the final instruction/nop can be broken
+
+`--n` Instead replace whole instruction(s), insert the number specified by ITEMS-value of nop(s)-instruction(s)
```
Comment (Grazfather) [inline docs/commands/nop.md:15]:
```suggestion
`--n` Instead of replacing whole instructions, insert ITEMS nop instructions, no matter how many instructions it overwrites
```

File: docs/commands/nop.md [hunk 3]
```diff
@@ -3,21 +3,26 @@
...
+
+`--b` Instead replace whole instruction(s), fill with nop(s)-instruction(s) the number specified by ITEMS-value bytes
```
Comment (Grazfather) [inline docs/commands/nop.md:17]:
```suggestion
`--b` Instead of replacing whole instructions, fill ITEMS bytes with nops
```

File: docs/commands/nop.md [hunk 4]
```diff
@@ -3,21 +3,26 @@
...
 ```
 
-`LOCATION` address/symbol to patch
+`LOCATION` address/symbol to patch (by default this command replace whole instruction(s))
 
-`--n NUM_ITEMS` Instead of writing one instruction/nop, patch the specified number of
-instructions/nops (full instruction size by default)
+`--i ITEMS` number of items to insert (default 1)
+
+`--f` Force patch when the final instruction/nop can be broken
```
Comment (Grazfather) [inline docs/commands/nop.md:13]:
```suggestion
`--f` Force patch even when the selected settings could overwrite partial instructions
```

File: gef.py [hunk 5]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
-                 "\n\tLOCATION\taddress/symbol to patch"
-                 "\t--n NUM_ITEMS\tInstead of writing one instruction/nop, patch the specified number of instructions/nops (full instruction size by default)"
-                 "\t--b\tInstead of writing full instruction size, patch the specified number of nops")
-    _example_ = f"{_cmdline_} $pc"
-
+    _syntax_  = ("{_cmdline_} [LOCATION] [--i ITEMS] [--f] [--n] [--b]"
+                 "\n\tLOCATION\taddress/symbol to patch (by default this command replace whole instruction(s))"
+                 "\t--i ITEMS\tnumber of items to insert (default 1)"
+                 "\t--f\tForce patch when the final instruction/nop can be broken"
+                 "\t--n\tInstead replace whole instruction(s), insert the number specified by ITEMS-value of nop(s)-instruction(s)"
```
Comment (Grazfather) [inline gef.py:6032]: take from my suggestions above

File: gef.py [hunk 6]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+                 f"{_cmdline_} --f --b --i 2 $pc+3"
+                 f"{_cmdline_} --n --i 2 $pc+3",]
 
     def __init__(self) -> None:
         super().__init__(complete=gdb.COMPLETE_LOCATION)
         return
 
     @only_if_gdb_running
-    @parse_arguments({"address": "$pc"}, {"--n": 0, "--b": False})
+    @parse_arguments({"address": "$pc"}, {"--i": 1, "--b": True, "--f": True, "--n": True})
```
Comment (Grazfather) [inline gef.py:6047]: Do we want force to be true? and how can b and n both be true?

File: gef.py [hunk 7]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
         nop = gef.arch.nop_insn
-        num_items = args.n or 1
-        as_nops_flags = not args.b
+        num_items = args.i or 1
+        fill_bytes = args.b 
+        fill_nops = args.n
+        force_flag = args.f or False
+        fill_instructions = False if fill_nops or fill_bytes else True
+   
+        if fill_nops + fill_bytes + fill_instructions != 1:
```
Comment (Grazfather) [inline gef.py:6058]:
```suggestion
        if fill_nops and fill_bytes:
```

File: gef.py [hunk 8]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+            err("only is possible specify --b or --n at same time")
```
Comment (Grazfather) [inline gef.py:6059]: you have to return here

File: gef.py [hunk 9]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
             except:
-                err(f"Cannot patch instruction at {address:#x}: MAYBE reaching unmapped area")
+                err(f"Cannot patch instruction at {address:#x} reaching unmapped area")
                 return
             total_bytes = (last_addr - address) + gef_get_instruction_at(last_addr).size()
 
-        if total_bytes % len(nop):
-            warn(f"Patching {total_bytes} bytes at {address:#x} will result in a partially patched instruction and may break disassembly")
+        if len(nop) > total_bytes or total_bytes % len(nop):
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-NOP (byte nr {total_bytes % len(nop):#x}) broken and may cause a crash or break disassembly")
```
Comment (Grazfather) [inline gef.py:6075]: please break up long lines

File: gef.py [hunk 10]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+            if not force_flag:
+                err("you must use --f to allow this kind of patch")
```
Comment (Grazfather) [inline gef.py:6077]:
```suggestion
                err("Use --f (force) to ignore this warning")
```

File: gef.py [hunk 11]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+            if not Address(value=curr_ins.address + 1).valid:
+                err(f"Cannot patch instruction at {address:#x}: reaching unmapped area")
+                return
+            curr_ins = gef_next_instruction(curr_ins.address)
 
-        nops = bytearray(nop * (total_bytes // len(nop)))
+        final_ins_end_addr = curr_ins.address + curr_ins.size()
+        
+        if final_ins_end_addr != target_end_address:
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-INSTRUCTION ({curr_ins.address:#x}) broken and may cause a crash or break disassembly")
```
Comment (Grazfather) [inline gef.py:6091]:
```suggestion
            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-INSTRUCTION ({curr_ins.address:#x}) being partial overwritten and may cause a crash or break disassembly")
```
and also wrap

File: gef.py [hunk 12]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
     @only_if_gdb_running
-    @parse_arguments({"address": "$pc"}, {"--n": 0, "--b": False})
+    @parse_arguments({"address": "$pc"}, {"--i": 1, "--b": True, "--f": True, "--n": True})
     def do_invoke(self, _: List[str], **kwargs: Any) -> None:
         args : argparse.Namespace = kwargs["arguments"]
         address = parse_address(args.address)
         nop = gef.arch.nop_insn
-        num_items = args.n or 1
-        as_nops_flags = not args.b
+        num_items = args.i or 1
```
Comment (hugsy) [inline gef.py:6052]: 1 is already the default value

```suggestion
        num_items = args.i
```

File: gef.py [hunk 13]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+                return
+            curr_ins = gef_next_instruction(curr_ins.address)
 
-        nops = bytearray(nop * (total_bytes // len(nop)))
+        final_ins_end_addr = curr_ins.address + curr_ins.size()
+        
+        if final_ins_end_addr != target_end_address:
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-INSTRUCTION ({curr_ins.address:#x}) broken and may cause a crash or break disassembly")
+            if not force_flag:
+                err("you must use --f to allow this kind of patch")
```
Comment (hugsy) [inline gef.py:6093]:
```suggestion
                err("You must use --f to allow misaligned patching.")
```

File: gef.py [hunk 14]
```diff
@@ -6025,54 +6025,83 @@ class NopCommand(GenericCommand):
...
+                return
+
+        nops = bytearray(nop * total_bytes) # this array will be bigger than needed when arch nop is > 1 but who cares
```
Comment (hugsy) [inline gef.py:6096]: Comment is unhelpful, remove

**Review by hugsy — CHANGES_REQUESTED** [review]
a few things to changes (also from graz's review) then we're good

File: docs/commands/nop.md (hunk 1 above)
Comment (therealdreg) [reply docs/commands/nop.md:9]: done

File: docs/commands/nop.md (hunk 2 above)
Comment (therealdreg) [reply docs/commands/nop.md:15]: done

File: docs/commands/nop.md (hunk 3 above)
Comment (therealdreg) [reply docs/commands/nop.md:17]: done

File: docs/commands/nop.md (hunk 4 above)
Comment (therealdreg) [reply docs/commands/nop.md:13]: done

File: gef.py (hunk 5 above)
Comment (therealdreg) [reply gef.py:6032]: done

File: gef.py (hunk 7 above)
Comment (therealdreg) [reply gef.py:6058]: done

File: gef.py (hunk 8 above)
Comment (therealdreg) [reply gef.py:6059]: done

File: gef.py (hunk 9 above)
Comment (therealdreg) [reply gef.py:6075]: done

File: gef.py (hunk 10 above)
Comment (therealdreg) [reply gef.py:6077]: done

File: gef.py (hunk 11 above)
Comment (therealdreg) [reply gef.py:6091]: done

File: gef.py (hunk 12 above)
Comment (therealdreg) [reply gef.py:6052]: if the user specify --i 0 I want force min 1

File: gef.py (hunk 13 above)
Comment (therealdreg) [reply gef.py:6093]: done

File: gef.py (hunk 14 above)
Comment (therealdreg) [reply gef.py:6096]: done

**Comment by therealdreg** [comment]
btw, @hugsy suggests in discord join the "warning msg" with "use --f msg": done

File: gef.py (hunk 6 above)
Comment (therealdreg) [reply gef.py:6047]: if the user specify --b and --n both will be true

File: docs/commands/nop.md [hunk 15]
```diff
@@ -3,21 +3,26 @@
...
+
+`--f` Force patch even when the selected settings could overwrite partial instructions
+
+`--n` Instead of replacing whole instructions, insert ITEMS nop instructions, no matter how many instructions it overwrites
+
+`--b` Instead of replacing whole instructions, fill ITEMS bytes with nops
 
-`--b` Instead of writing full instruction size, patch the specified number of nops
 
 ```bash
```
Comment (hugsy) [inline docs/commands/nop.md:20]: This command has become a bit convoluted. Can you also add to the example what the commands do in plain english?

**Review by hugsy — CHANGES_REQUESTED** [review]
I think examples in the docs need a bit more explanation in plain english.

File: docs/commands/nop.md (hunk 15 above)
Comment (therealdreg) [reply docs/commands/nop.md:20]: me?? in english??? xDD oka, I will try

Comment (therealdreg) [reply docs/commands/nop.md:20]: done

**Review by hugsy — APPROVED** [review]

File: docs/commands/nop.md [hunk 16]
```diff
@@ -3,21 +3,50 @@
...
+`LOCATION` address/symbol to patch (by default this command replaces whole instructions)
 
-`--n NUM_ITEMS` Instead of writing one instruction/nop, patch the specified number of
-instructions/nops (full instruction size by default)
+`--i ITEMS` number of items to insert (default 1)
 
-`--b` Instead of writing full instruction size, patch the specified number of nops
+`--f` Force patch even when the selected settings could overwrite partial instructions
 
+`--n` Instead of replacing whole instructions, insert ITEMS nop instructions, no matter how many instructions it overwrites
```
Comment (Grazfather) [inline docs/commands/nop.md:15]: wrap to 100 according to editorconfig

File: docs/commands/nop.md [hunk 17]
```diff
@@ -3,21 +3,50 @@
...
+```bash
+gef➤ 	nop --i 2 $pc+3
+```
+
+Replace 1 byte with nop at current instruction ($pc):
+```bash
 gef➤ 	nop --b
+```
+
+Replace 1 byte with nop at $pc+3 address:
```
Comment (Grazfather) [inline docs/commands/nop.md:39]:
```suggestion
Replace 1 byte with nop at address $pc+3:
```

File: docs/commands/nop.md [hunk 18]
```diff
@@ -3,21 +3,50 @@
...
+```bash
 gef➤ 	nop --b $pc+3
-gef➤ 	nop --b --n 2 $pc+3
+```
+
+Replace 2 bytes with nop(s) (breaking the last instruction) at $pc+3 address:
```
Comment (Grazfather) [inline docs/commands/nop.md:44]:
```suggestion
Replace 2 bytes with nop(s) (breaking the last instruction) at address $pc+3:
```

File: docs/commands/nop.md [hunk 19]
```diff
@@ -3,21 +3,50 @@
...
+```bash
+gef➤ 	nop --f --b --i 2 $pc+3
+```
+
+Patching with 2 nops at $pc+3 address:
```
Comment (Grazfather) [inline docs/commands/nop.md:49]: Patch 2 nops at address $pc+3

File: gef.py (hunk 6 above)
Comment (Grazfather) [reply gef.py:6047]: but how can that work?

File: gef.py [hunk 20]
```diff
@@ -6025,54 +6025,85 @@ class NopCommand(GenericCommand):
...
                 return
             total_bytes = (last_addr - address) + gef_get_instruction_at(last_addr).size()
 
-        if total_bytes % len(nop):
-            warn(f"Patching {total_bytes} bytes at {address:#x} will result in a partially patched instruction and may break disassembly")
+        if len(nop) > total_bytes or total_bytes % len(nop):
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-NOP "
+                 f"(byte nr {total_bytes % len(nop):#x}) broken and may cause a crash or break disassembly. "
+                 f"Use --f (force) to ignore this warning")
+            if not force_flag:
```
Comment (Grazfather) [inline gef.py:6078]: I would warn even if we force, but don't explain about --f , so I would add the `Use ...` part in this case

File: gef.py [hunk 21]
```diff
@@ -6025,54 +6025,85 @@ class NopCommand(GenericCommand):
...
+                return
```
Comment (Grazfather) [inline gef.py:6079]: Here err(Use --f...

File: gef.py [hunk 22]
```diff
@@ -6025,54 +6025,85 @@ class NopCommand(GenericCommand):
...
+
+        target_end_address = address + total_bytes
+        curr_ins = gef_current_instruction(address)
+        while curr_ins.address + curr_ins.size() < target_end_address:
+            if not Address(value=curr_ins.address + 1).valid:
```
Comment (Grazfather) [inline gef.py:6084]: the +1 is confusing. Why not but this check after `curr_ins = get next` so you don't need the +1?

isn't `curr_ins.address` already an Address? if so, `if not curr_ins.valid`

File: gef.py [hunk 23]
```diff
@@ -6025,54 +6025,85 @@ class NopCommand(GenericCommand):
...
+            curr_ins = gef_next_instruction(curr_ins.address)
 
-        nops = bytearray(nop * (total_bytes // len(nop)))
+        final_ins_end_addr = curr_ins.address + curr_ins.size()
+        
+        if final_ins_end_addr != target_end_address:
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-INSTRUCTION "
+                 f"({curr_ins.address:#x}) being partial overwritten and may cause a crash or break disassembly. "
+                 f"You must use --f to allow misaligned patching.")
+            if not force_flag:
```
Comment (Grazfather) [inline gef.py:6095]: Again, put the -f part here f"Use --f (force) to ignore this warning"

File: gef.py [hunk 24]
```diff
@@ -6025,54 +6025,85 @@ class NopCommand(GenericCommand):
...
+                return
+            curr_ins = gef_next_instruction(curr_ins.address)
 
-        nops = bytearray(nop * (total_bytes // len(nop)))
+        final_ins_end_addr = curr_ins.address + curr_ins.size()
+        
+        if final_ins_end_addr != target_end_address:
+            warn(f"Patching {total_bytes} bytes at {address:#x} will result in LAST-INSTRUCTION "
+                 f"({curr_ins.address:#x}) being partial overwritten and may cause a crash or break disassembly. "
+                 f"You must use --f to allow misaligned patching.")
```
Comment (Grazfather) [inline gef.py:6094]: No sense printing this if they already did --f

Comment (therealdreg) [reply gef.py:6094]: @hugsy suggests on discord make only ONE msg (my code before was like you suggest)

File: gef.py (hunk 23 above)
Comment (therealdreg) [reply gef.py:6095]: @hugsy suggests on discord make only ONE msg (my code before was like you suggest)

File: gef.py (hunk 22 above)
Comment (therealdreg) [reply gef.py:6084]: because I dont want disasm invalid Address (avoid exception / error .. etc) I think the current code is OK

File: gef.py (hunk 21 above)
Comment (therealdreg) [reply gef.py:6079]: @hugsy suggests on discord make only ONE msg (my code before was like you suggest)

File: docs/commands/nop.md (hunk 17 above)
Comment (therealdreg) [reply docs/commands/nop.md]: done

File: gef.py (hunk 20 above)
Comment (therealdreg) [reply gef.py:6078]: @hugsy suggests on discord make only ONE msg (my code before was like you suggest)

File: docs/commands/nop.md (hunk 16 above)
Comment (therealdreg) [reply docs/commands/nop.md:15]: done

**Comment by therealdreg** [comment]
added a test to check if --b and --n at same time == error

File: docs/commands/nop.md (hunk 16 above)
Comment (Grazfather) [reply docs/commands/nop.md:15]: I don't think it is?

**Review by hugsy — APPROVED** [review]
